### PR TITLE
WIP: Support device public key and passkeys

### LINKED ISSUE #334: Library lacks support for passkeys and devicePublicKey extension
This library currently lacks support for [multi-device FIDO credentials](https://w3c.github.io/webauthn/#backup-eligibility) (aka [passkeys](https://passkeys.dev/)) as well as the related [devicePublicKey ](https://w3c.github.io/webauthn/#sctn-device-publickey-extension)extension which lets passkeys not be a deal-breaker for [enterprise](https://w3c.github.io/webauthn/#dom-attestationconveyancepreference-enterprise) customers.  I should fix this.

### PR DESCRIPTION
**Pull request by aseigler**
Resolves #334 (and #350).

### CODE REVIEW

File: Src/Fido2.Models/Objects/AuthenticationExtensionsDevicePublicKeyInputs.cs
Comment (iamcarbon) [inline Src/Fido2.Models/Objects/AuthenticationExtensionsDevicePublicKeyInputs.cs:4]: Use a file scoped namespace?

File: Src/Fido2.Models/Objects/AuthenticationExtensionsDevicePublicKeyOutputs.cs
Comment (iamcarbon) [inline Src/Fido2.Models/Objects/AuthenticationExtensionsDevicePublicKeyOutputs.cs:3]: Use a file scoped namespace?

File: Src/Fido2/Objects/AuthenticatorData.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/AuthenticatorData.cs:51]: Consider renaming IsBackupEligible

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:6]: Use file-scoped namespace.

File: Src/Fido2/Objects/AttestedCredentialData.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/AttestedCredentialData.cs:19]: change to const int?

File: Src/Fido2/Objects/AttestedCredentialData.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/AttestedCredentialData.cs:18]: change to const int _minLength = 12; // guid (8) + short (2) + byte (1) + byte (1) ?

File: Src/Fido2/DevelopmentInMemoryStore.cs
Comment (iamcarbon) [inline Src/Fido2/DevelopmentInMemoryStore.cs:78]: Add a return after each property for readability?

File: Src/Fido2/AuthenticatorAssertionResponse.cs
Comment (iamcarbon) [inline Src/Fido2/AuthenticatorAssertionResponse.cs:383]: Throw FidoVerificationException with inner exception?

File: Src/Fido2.Models/Objects/PublicKeyCredentialType.cs
Comment (iamcarbon) [inline Src/Fido2.Models/Objects/PublicKeyCredentialType.cs:15]: Add line break for readability.

File: Src/Fido2.Models/Objects/AssertionVerificationResult.cs
Comment (iamcarbon) [inline Src/Fido2.Models/Objects/AssertionVerificationResult.cs:16]: Add line break for readability.

**Comment by iamcarbon** [comment]
Hi @aseigler -- anything left here?

File: Src/Fido2/Objects/AttestedCredentialData.cs
Comment (aseigler) [reply Src/Fido2/Objects/AttestedCredentialData.cs:18]: Eh, I kind of like this one the way it is, matches the sizes mentioned in the spec.  Guid is 16 bytes, not 8.

File: Src/Fido2/AuthenticatorAssertionResponse.cs
Comment (aseigler) [reply Src/Fido2/AuthenticatorAssertionResponse.cs:383]: Yeah, I meant to go back and figure out what to throw for each of these but hadn't yet...and now I have to revisit again as some things have changed.

**Comment by aseigler** [comment]
> Hi @aseigler -- anything functionality left here?

Not sure, need to do another pass on the DPK things to make sure I've got all the logic straight.  The passkey area I _think_ is good to go.

File: Src/Fido2/Objects/AttestedCredentialData.cs
Comment (iamcarbon) [reply Src/Fido2/Objects/AttestedCredentialData.cs:16]: Ugh -- yes, Guid = 16. Mostly wanted to kill the call and make this a const, but given my own mistake above, maybe best as is! :)

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:21]: Do we need to consider the Endianness of the AaGuid when building the Authentication Matcher?

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:25]: Is aaguid BigEndian here?

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:43]: Can we remove the private setter (making the backing fields readonly)

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:50]: Remove private setter

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:67]: Consider backing with a ```private readonly byte[] _nonce```, and exposing over ```ReadOnlySpan<byte> Nonce => _nonce```; to prevent mutation.

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:60]: Remove private setter

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:84]: Remove private setter.

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (iamcarbon) [inline Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:11]: make internal static to avoid allocations per instance... or use compiler optimized ReadOnlySpan<byte> trick.

```sharp
internal static ReadOnlySpan<byte> _dpkAuthDataPrefix => new byte[] {
   100, 101, 118, 105, 99, 101, 32, 98, 111, 117, 110, 100, 32, 107, 101, 121, 32, 97, 116, 
   116, 101, 115, 116, 97, 116, 105, 111, 110, 32, 115, 105, 103, 0, 255, 255, 255, 255
 };
 ```

**Review by iamcarbon — CHANGES_REQUESTED** [review]

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (aseigler) [reply Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:25]: Probably.  I'll add the same logic as we use elsewhere.

File: Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs
Comment (aseigler) [reply Src/Fido2/Objects/DevicePublicKeyAuthenticatorOutput.cs:21]: Not as long as we're consistent with the storage during registration vs logon, which we will be.

File: Src/Fido2/Objects/AttestedCredentialData.cs
Comment (aseigler) [reply Src/Fido2/Objects/AttestedCredentialData.cs:16]: I am just going to go with a `readonly int` value of 20.  This isn't spec level check, it's a sanity check added so we can tell when the conformance test that sends a value that is way too small we can tell that is specifically the test we passed and not some other problem.

**Comment by aseigler** [comment]
Still need to make RP configurable BE/BS options and plumb those into registration and login flows before passkey is done.

DPK looks like it has object names that have been updated. 

Reminder to create CredentialRecord object, move existing StoredCredential members there and hand around that credential record instead of parts of it in and out of login flow.

Reminder to remove the rest of the references to token binding.

File: Demo/Startup.cs
Comment (rudiv) [inline Demo/Startup.cs:55]: I would argue that these be allowed by default (ie. `Configuration.GetValue<bool>("fido2:allowBackupEligibleCredential", true);`) to support a happier migration path. Some MPs may not care about these statuses at all, and so their existing configuration would break?

Comment (aseigler) [reply Demo/Startup.cs:55]: > I would argue that these be allowed by default to support a happier upgrade path. Some RPs may not care about these statuses at all, and so their existing configuration would break?
> 
> I know this comment is against the Demo project, but the [Configuration default](https://github.com/aseigler/fido2-net-lib/blob/feat-passkey/Src/Fido2.Models/Fido2Configuration.cs#L120) would also be false if this wasn't decided / set at startup.

I hear ya.  This is a enterprise vs consumer setting in my eyes, and I personally am enterprise focused.  When this PR is in a release it will be part of a major update, and we are expecting that will contain some breaking changes.  Regardless of which way it goes by default, there will be notes to help guide implementers.

Comment (rudiv) [reply Demo/Startup.cs:55]: I agree, I think my main concern is that this may block typical use cases.

For an example (even thinking enterprise), you may wish for your users to have a passkey that's backed up and set this setting. However, they would then be unable to add a physical key (Yubikey et al). I know this is the whole point of "Passkeys" to not need it, but I feel that it's more of a per-key decision than one that should be set globally for the RP?

Comment (abergs) [reply Demo/Startup.cs:55]: @aseigler I agree that we should default to true. Given that the average developer might not know what this option do (and should be True for them), while I think we can expect enterprise developers have more knowledge and know to turn it off because of their own research/policy for deploying fido2.

**Comment by iamcarbon** [comment]
@aseigler Anything else left here before we merge?

**Comment by aseigler** [comment]
> @aseigler Anything else left here before we merge?

I think the only thing that hasn't been gotten to in the list was moving StoredCredential to CredentialRecord but that can be done separately, no need to hold this up for that.

File: Demo/Startup.cs
Comment (aseigler) [reply Demo/Startup.cs:55]: > 
They would actually be able to register a typical security key with this setting true or false, as those do not set BE flag. Currently only certain platform authenticators do. It's undefined how the platform is protecting those BE keys and where they are going to be restored to. That's why in a situation where the RP cares (enterprise), they might wish to restrict BE keys from being registered at all on the platform, or maybe allow some blessed authenticators but not others (would require additional work to support this). I guess there might be a reason an RP would only want to allow BE keys? That's not the result or intent of this implementation, at least currently.

Comment (rudiv) [reply Demo/Startup.cs:55]: Understand the Enterprise point of view.

From a more consumer-focused point of view (ie. specifically the well marketed "Passkeys" aspect) we may want to accept _only_ Passkeys at first, to ensure that users are not able to lose access very easily. Whilst it's not perfect and most authenticators don't give on where the keys are from / stored, knowing that they're backed up allows it to be reasonably assumed that the key is durable and isn't going to disappear like in the case of a physical key loss. For example, we explicitly advise the user to register with a backed up authenticator (specifically iCloud backed Passkeys or their Android phone).

I might be wrong in assuming that this is where this library is going to see most of its use in the short-mid term, but that's my 2c on why this default would be the wrong one.

**Comment by Dannzak** [comment]
Hey! Any news on this?

File: Src/Fido2/AuthenticatorAttestationResponse.cs
Comment (abergs) [inline Src/Fido2/AuthenticatorAttestationResponse.cs:213]: Why do we not return Transports anymore?

File: Src/Fido2/DevelopmentInMemoryStore.cs
Comment (abergs) [inline Src/Fido2/DevelopmentInMemoryStore.cs:125]: Should we depricate the descriptor or do we still have a use for it? (Now that we've added Id, PublicKey, Type to the class itself)

File: Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs
Comment (abergs) [inline Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs:24]: Do we want to use the same JsonConverter as we do on PublicKey?

File: Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs
Comment (abergs) [inline Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs:42]: Do we want to use JsonConverter on this and `AttestationClientDataJSON`?

**Review by abergs — APPROVED** [review]
Great work on these two large features. I have some comments, but I'm happy with the changes.

File: Src/Fido2/AuthenticatorAttestationResponse.cs
Comment (aseigler) [reply Src/Fido2/AuthenticatorAttestationResponse.cs:213]: We didn't return transports before, in fact we didn't even collect it in the JavaScript at all, I added in [this](https://github.com/passwordless-lib/fido2-net-lib/blob/53caf81da894691ad10eefcce5fd5979b4a8e6ed/Demo/wwwroot/js/custom.register.js#L129) but it looks like it needs to be added to all of the other registration JavaScript files, and collect transports in [AuthenticatorAttestationRawResponse.ResponseData](https://github.com/passwordless-lib/fido2-net-lib/blob/53caf81da894691ad10eefcce5fd5979b4a8e6ed/Src/Fido2.Models/AuthenticatorAttestationRawResponse.cs#L26), and then populate transports in the [controller from the raw response](https://github.com/passwordless-lib/fido2-net-lib/blob/53caf81da894691ad10eefcce5fd5979b4a8e6ed/Demo/Controller.cs#L131).  There are a few reasons to collect transports, to help with user facing dialogs, but also to compare what transports are being used against a policy, or comparing what transports should be available according to the authenticator metadata.  Basically this is incomplete at the moment.

File: Demo/TestController.cs
Comment (abergs) [reply Demo/TestController.cs:183]: Hmm... I think I understand @aseigler. Am I right to think that before saving the DPK the RP would perhaps check some policy or trigger a risk based ceremony before trusting that DPK and then saving/updating the storage with `res.DevicePublicKey` but in our demo we just add it directly to your in memory storage?

Comment (aseigler) [reply Demo/TestController.cs:183]: Yeah, basically same as registering a new credential against an existing user, but _way_ more complicated.

Comment (abergs) [reply Demo/TestController.cs:183]: Alright makes sense, I was just confused about how we used it in our demo. When I get the time I'll add a `// TODO: Check policy if we allow new devices, if-so, store it to db`

File: Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs
Comment (aseigler) [reply Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs:24]: Already done in AuthenticatorAttestationRawResponse, no?

File: Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs
Comment (aseigler) [reply Src/Fido2.Models/Objects/AttestationVerificationSuccess.cs:42]: Already done in AuthenticatorAttestationRawResponse, no?

**Pull request merged**
